code/modules/reagents/chemistry/equilibrium.dm

```diff
@@ -0,0 +1,226 @@
+		for(var/P in reaction.required_catalysts)
+			var/datum/reagent/R0 = P
+			if(R0 == R.type)
+				total_matching_catalysts++
+
+		for(var/B in reaction.required_reagents)
+			var/datum/reagent/R0 = B
+			if(R0 == R.type) // required_reagents = list(/datum/reagent/consumable/sugar = 1) /datum/reagent/consumable/sugar
+				total_matching_reagents++
+	
+	if(!(total_matching_reagents == reaction.required_reagents.len))
+		return FALSE
+
+	if(!(total_matching_catalysts == reaction.required_catalysts.len))
+		return FALSE
+
+	//All good!
+	return TRUE
+
+//Calculates how much we're aiming to create
+/datum/equilibrium/proc/calculate_yield()
+	if(toDelete)
+		return FALSE
+	if(!reaction)
+		WARNING("Tried to calculate an equlibrium for reaction [reaction.type], but there was no reaction set for the datum")
+		return FALSE
+	multiplier = INFINITY
+	productRatio = 0
+	targetVol = 0
+	for(var/B in reaction.required_reagents)
+		multiplier = min(multiplier, round((holder.get_reagent_amount(B) / reaction.required_reagents[B]), CHEMICAL_QUANTISATION_LEVEL))
+	for(var/P in reaction.results)
+		targetVol += (reaction.results[P]*multiplier)
+		productRatio += reaction.results[P]
+	if(targetVol == 0 || multiplier == INFINITY)
+		return FALSE
+	return TRUE
+
+//Main reaction processor
+//Increments reaction by a timestep
+/datum/equilibrium/proc/react_timestep(delta_time)
+	if(!check_conditions())
+		toDelete = TRUE
+		return
+	if(!calculate_yield())
+		toDelete = TRUE
+		return
+
+
+	var/deltaT = 0 //how far off optimal temp we care
+	var/deltapH = 0 //How far off the pH we are
+	var/cached_pH = holder.pH
+	var/cached_temp = holder.chem_temp
+	var/purity = 1 //purity of the current step
+
+	//Begin checks
+	//Calculate DeltapH (Deviation of pH from optimal)
+	//Within mid range
+	if (cached_pH >= reaction.optimal_pH_min  && cached_pH <= reaction.optimal_pH_max)
+		deltapH = 1
+	//Lower range
+	else if (cached_pH < reaction.optimal_pH_min)
+		if (cached_pH < (reaction.optimal_pH_min - reaction.determin_pH_range))
+			deltapH = 0
+			//If outside pH range, 0
+		else
+			deltapH = (((cached_pH - (reaction.optimal_pH_min - reaction.determin_pH_range))**reaction.pH_exponent_factor)/((reaction.determin_pH_range**reaction.pH_exponent_factor))) //main pH calculation
+	//Upper range
+	else if (cached_pH > reaction.optimal_pH_max)
+		if (cached_pH > (reaction.optimal_pH_max + reaction.determin_pH_range))
+			deltapH = 0
+			//If outside pH range, 0
+		else
+			deltapH = (((- cached_pH + (reaction.optimal_pH_max + reaction.determin_pH_range))**reaction.pH_exponent_factor)/(reaction.determin_pH_range**reaction.pH_exponent_factor))//Reverse - to + to prevent math operation failures.
+	
+	//This should never proc, but it's a catch incase someone puts in incorrect values
+	else
+		WARNING("[holder.my_atom] attempted to determine FermiChem pH for '[reaction.type]' which had an invalid pH of [cached_pH] for set recipie pH vars. It's likely the recipe vars are wrong.")
```

Comment:
```suggestion
		stack_trace("[holder.my_atom] attempted to determine FermiChem pH for '[reaction.type]' which had an invalid pH of [cached_pH] for set recipie pH vars. It's likely the recipe vars are wrong.")
```
Favor `CRASH()` and `stack_trace()` over `WARNING`, as in live servers those compile and store bug information better.

Reply:
Thanks for the correction. I'll keep it in mind going forward!